sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1971,65 +2025,84 @@ class Analyzer(
    */
   object ResolveFunctions extends Rule[LogicalPlan] {
     val trimWarningEnabled = new AtomicBoolean(true)
+
+    def resolveFunction(): PartialFunction[Expression, Expression] = {
+      case u if !u.childrenResolved => u // Skip until children are resolved.
+      case u: UnresolvedAttribute if resolver(u.name, VirtualColumn.hiveGroupingIdName) =>
+        withPosition(u) {
+          Alias(GroupingID(Nil), VirtualColumn.hiveGroupingIdName)()
+        }
+      case u@UnresolvedGenerator(name, children) =>
+        withPosition(u) {
+          v1SessionCatalog.lookupFunction(name, children) match {
+            case generator: Generator => generator
+            case other =>
+              failAnalysis(s"$name is expected to be a generator. However, " +
+                s"its class is ${other.getClass.getCanonicalName}, which is not a generator.")
+          }
+        }
+      case u@UnresolvedFunction(funcId, arguments, isDistinct, filter) =>
+        withPosition(u) {
+          v1SessionCatalog.lookupFunction(funcId, arguments) match {
+            // AggregateWindowFunctions are AggregateFunctions that can only be evaluated within
+            // the context of a Window clause. They do not need to be wrapped in an
+            // AggregateExpression.
+            case wf: AggregateWindowFunction =>
+              if (isDistinct || filter.isDefined) {
+                failAnalysis("DISTINCT or FILTER specified, " +
+                  s"but ${wf.prettyName} is not an aggregate function")
+              } else {
+                wf
+              }
+            // We get an aggregate function, we need to wrap it in an AggregateExpression.
+            case agg: AggregateFunction =>
+              if (filter.isDefined && !filter.get.deterministic) {
+                failAnalysis("FILTER expression is non-deterministic, " +
+                  "it cannot be used in aggregate functions")
+              }
+              AggregateExpression(agg, Complete, isDistinct, filter)
+            // This function is not an aggregate function, just return the resolved one.
+            case other if (isDistinct || filter.isDefined) =>
+              failAnalysis("DISTINCT or FILTER specified, " +
+                s"but ${other.prettyName} is not an aggregate function")
+            case e: String2TrimExpression if arguments.size == 2 =>
+              if (trimWarningEnabled.get) {
+                log.warn("Two-parameter TRIM/LTRIM/RTRIM function signatures are deprecated." +
```

Comment:
(I know this is not related to this PR though) `logWarning`?

Reply:
Done